src/pudl/analysis/allocate_net_gen.py

```diff
@@ -731,6 +1028,121 @@ def allocate_net_gen_by_gen_esc(gen_pm_fuel):
+    but all will end up with a ``frac`` column that can be used to allocate
+    the ``fuel_consumed_mmbtu_gf_tbl``.
+
+    Args:
+        gen_pm_fuel (pandas.DataFrame): output of :func:``prep_alloction_fraction()``.
+
+    """
+    # break out the table into these four different generator types.
+    all_gen = gen_pm_fuel.loc[gen_pm_fuel.in_bf_tbl_all]
+    some_gen = gen_pm_fuel.loc[gen_pm_fuel.in_bf_tbl_any & ~gen_pm_fuel.in_bf_tbl_all]
+    bf_only = gen_pm_fuel.loc[~gen_pm_fuel.in_bf_tbl_any]
+
+    logger.info(
+        "Ratio calc types: \n"
+        f"   All gens w/in generation table:  {len(all_gen)}#, {all_gen.capacity_mw.sum():.2} MW\n"
+        f"   Some gens w/in generation table: {len(some_gen)}#, {some_gen.capacity_mw.sum():.2} MW\n"
+        f"   No gens w/in generation table:   {len(bf_only)}#, {bf_only.capacity_mw.sum():.2} MW"
+    )
+    if len(gen_pm_fuel) != len(all_gen) + len(some_gen) + len(bf_only):
+        raise AssertionError(
+            "Error in splitting the gens between records showing up fully, "
+            "partially, or not at all in the generation table."
+        )
+
+    # In the case where we have all of the fuel from the bf
+    # table, we still allocate, because the fuel reported in these two
+    # tables don't always match perfectly
+    all_gen = all_gen.assign(
+        frac_fuel=lambda x: x.fuel_consumed_mmbtu_bf_tbl
+        / x.fuel_consumed_mmbtu_bf_tbl_pm_fuel,
+        frac=lambda x: x.frac_fuel,
+    )
+    # _ = _test_frac(all_gen)
+
+    some_gen = some_gen.assign(
+        # fraction of the fuel consumption that should go to the generators that
+        # report in the boiler fuel table
+        frac_from_bf_tbl=lambda x: x.fuel_consumed_mmbtu_bf_tbl_pm_fuel
+        / x.fuel_consumed_mmbtu_gf_tbl,
+        # for records within these mix groups that do have fuel consumption in the
+        # bf table..
+        frac_fuel=lambda x: x.fuel_consumed_mmbtu_bf_tbl
+        / x.fuel_consumed_mmbtu_bf_tbl_pm_fuel,  # generator based fuel from bf table
+        frac_bf=lambda x: x.frac_fuel * x.frac_from_bf_tbl,
+        # fraction of fuel that does not show up in the bf table
+        # set minimum fraction to zero so we don't get negative fuel
+        frac_missing_from_bf_tbl=lambda x: np.where(
+            (x.frac_from_bf_tbl < 1), (1 - x.frac_from_bf_tbl), 0
+        ),
+        capacity_mw_missing_from_bf_tbl=lambda x: np.where(
+            x.in_bf_tbl, 0, x.capacity_mw
+        ),
+        frac_cap=lambda x: x.frac_missing_from_bf_tbl
+        * (x.capacity_mw_missing_from_bf_tbl / x.capacity_mw_fuel_in_bf_tbl_group),
+        # the real deal
+        # this could aslo be `x.frac_bf + x.frac_cap` because the frac_bf
+        # should be 0 for any generator that does not have fuel in the bf_tbl
+        # and frac_cap should be 0 for any generator that has fuel in the
+        # bf_tbl.
+        frac=lambda x: np.where(x.in_bf_tbl, x.frac_bf, x.frac_cap),
```

Comment:
even after writing the analogous version of this for net gen, this still feels complicated but I think i've wrapped my head around it and this looks good! I think it would be good to add baby unit tests to demonstrate outcomes for this. If you want I could add some for the net get side of things (which i should have done originally!) as an example/template.